More dynamicDowncast<> adoption in editing code

annevk opened the pull request:
#### 31313f2c893f34e7e345f815f1a4ccdc9b6b5024
<pre>
More dynamicDowncast&lt;&gt; adoption in editing code
<a href="https://bugs.webkit.org/show_bug.cgi?id=268811">https://bugs.webkit.org/show_bug.cgi?id=268811</a>

Reviewed by Chris Dumez.

* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::targetFrame):
(WebCore::verticalScrollDistance):
* Source/WebCore/editing/FormatBlockCommand.cpp:
(WebCore::isElementForFormatBlock):
(WebCore::FormatBlockCommand::elementForFormatBlockCommand):
(WebCore::enclosingBlockToSplitTreeTo):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::removingNodeRemovesPosition):
(WebCore::CaretBase::paintCaret const):
(WebCore::FrameSelection::selectAll):
(WebCore::scanForForm):
* Source/WebCore/editing/InsertListCommand.cpp:
(WebCore::InsertListCommand::mergeWithNeighboringLists):
* Source/WebCore/editing/SimplifyMarkupCommand.cpp:
(WebCore::SimplifyMarkupCommand::doApply):
* Source/WebCore/editing/SpellChecker.cpp:
(WebCore::SpellChecker::isCheckable const):
* Source/WebCore/editing/SplitTextNodeContainingElementCommand.cpp:
(WebCore::SplitTextNodeContainingElementCommand::doApply):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::isInPasswordField const):

Canonical link: <a href="https://commits.webkit.org/274221@main">https://commits.webkit.org/274221@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61a32ef70681f3100378ff11d5c153d1628f06bf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40680 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40908 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40601 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20061 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14632 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38924 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12698 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/34275 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42187 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34688 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38518 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13267 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36719 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14829 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8624 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14293 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/editing/EditorCommand.cpp:
```diff
@@ -95,9 +95,10 @@ static LocalFrame* targetFrame(LocalFrame& frame, Event* event)
     if (!event)
         return &frame;
-    if (!is<Node>(event->target()))
+    RefPtr node = dynamicDowncast<Node>(event->target());
```
I don't think we need a RefPtr here. We're just calling a trivial getter.

cdumez commented on Source/WebCore/editing/InsertListCommand.cpp:
```diff
@@ -80,14 +80,13 @@ Ref<HTMLElement> InsertListCommand::mergeWithNeighboringLists(HTMLElement& list)
-        return nextList;
+    if (canMergeLists(&list, sibling.get())) {
+        mergeIdenticalElements(list, *sibling);
+        return *sibling;
```
This is causing unnecessary ref-counting churn. This should be `sibling.releaseNonNull()`

cdumez commented on Source/WebCore/editing/SpellChecker.cpp:
```diff
@@ -152,8 +152,8 @@ bool SpellChecker::isCheckable(const SimpleRange& range) const
         return false;
-    Ref node = range.start.container.get();
-    return !is<Element>(node) || downcast<Element>(node.get()).isSpellCheckingEnabled();
+    RefPtr element = dynamicDowncast<Element>(range.start.container.get());
```
I didn't look into `isSpellCheckingEnabled` but I doubt we really need a smart pointer here?

annevk replied:
It does node traversal (loop over parentOrShadowHostElement) and attribute lookup. You added the smart pointer 4 months ago as part of https://github.com/WebKit/WebKit/commit/389241b67619b49fab63ca2d9abafb5c2a69358a so I didn't want to regress that. Please confirm you definitely want to see it gone.

cdumez replied:
`isSpellCheckingEnabled` is not super trivial. Let's keep the smart pointer.

cdumez commented on Source/WebCore/editing/VisibleSelection.cpp:
```diff
@@ -683,8 +683,8 @@ Node* VisibleSelection::nonBoundaryShadowTreeRootNode() const
 {
-    RefPtr textControl = enclosingTextFormControl(start());
-    return is<HTMLInputElement>(textControl) && downcast<HTMLInputElement>(*textControl).isPasswordField();
+    RefPtr textControl = dynamicDowncast<HTMLInputElement>(enclosingTextFormControl(start()));
```
I didn't look into `isPasswordField()` but I doubt we need a smart pointer here.

annevk replied:
This was also added as part of 389241b67619b49fab63ca2d9abafb5c2a69358a. `isPasswordField()` is a somewhat trivial getter (although `HTMLInputElement` does delegate it to `InputType`). Happy to remove if you confirm.

Maybe that suggests that commit needs another review given the lessons learned in the past 4 months?

cdumez replied:
If it's not a straightforward data member getter, I think it is fine to keep a smart pointer.

cdumez commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -546,10 +546,11 @@ static bool removingNodeRemovesPosition(Node& node, const Position& position)
         return false;
 
-    return downcast<Element>(node).containsIncludingShadowDOM(position.anchorNode());
+    return element->containsIncludingShadowDOM(position.anchorNode());
```
I would have dropped the early return above and done:
```
return element && element->containsIncludingShadowDOM(position.anchorNode());
```

annevk committed:
More dynamicDowncast<> adoption in editing code
https://bugs.webkit.org/show_bug.cgi?id=268811

Reviewed by Chris Dumez.

* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::targetFrame):
(WebCore::verticalScrollDistance):
* Source/WebCore/editing/FormatBlockCommand.cpp:
(WebCore::isElementForFormatBlock):
(WebCore::FormatBlockCommand::elementForFormatBlockCommand):
(WebCore::enclosingBlockToSplitTreeTo):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::removingNodeRemovesPosition):
(WebCore::CaretBase::paintCaret const):
(WebCore::FrameSelection::selectAll):
(WebCore::scanForForm):
* Source/WebCore/editing/InsertListCommand.cpp:
(WebCore::InsertListCommand::mergeWithNeighboringLists):
* Source/WebCore/editing/SimplifyMarkupCommand.cpp:
(WebCore::SimplifyMarkupCommand::doApply):
* Source/WebCore/editing/SpellChecker.cpp:
(WebCore::SpellChecker::isCheckable const):
* Source/WebCore/editing/SplitTextNodeContainingElementCommand.cpp:
(WebCore::SplitTextNodeContainingElementCommand::doApply):
* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::isInPasswordField const):

Canonical link: https://commits.webkit.org/274221@main
